### DISCUSSION TITLE
Moving mongodb support back into preview / early access

### DISCUSSION THREAD

**Discussion by macrozone** [prisma/prisma#23888] [Q&A] [answered]
## Problem

Mongodb has been put Generally Available (GA) a while ago, 
this indicates that its suitable for production.

But there are so severe performance problems with that it may be devastating for the stability of your application if you use it in production.

https://github.com/prisma/prisma/issues/14871
https://github.com/prisma/prisma/issues/21453
https://github.com/prisma/prisma/issues/22812

and many more.

I and probably many more love prisma and were happy that it supports mongodb as well. However there is no indication that such severe problems exists. Mongodb beeing in Generally Available (GA) indicates a maturity level that it does not have. 

## Suggested solution

While I am currently stuck and can't switch easily to something else like mongose, we should at least avoid that new developers use prisma for mongodb for productive apps. So its best to flag it realistically as either preview or early-access and highlight the current limitations it has on the documentation

https://www.prisma.io/docs/orm/more/releases#early-access

Also its important to inform mongodb itself about this as they also promote prisma on their official website: https://cloud.mongodb.com/ecosystem/prisma

## Alternatives

Of course we all would be much happier if those problems would have been fixed in the first place. Some of them were known for a long time and well documented.

It has been made clear by the prisma team that mongodb has no priority. It was not tackled here https://prismaio.notion.site/Optimize-Prisma-Client-Query-Performance-68caf089581e4d07ac154dd34ad73639
and its not even on the roadmap https://prismaio.notion.site/Prisma-ORM-Roadmap-50766227b779464ab98899accb98295f

The team and the business defines the priority, that is fine. But please be realistic about what it can do and what it can't.

## Additional context

On a personal level i highly recommend to fix the underlying issues ASAP, because I think some would be rather trivial to fix. The damage this does to the reputation of both prisma and mongodb should not be underestimated.

**Comment by macrozone** [comment]
thx for migrating to discussion, I think its the better place. And seeing at least some reactions shows me that i am not alone.

just to be clear: I don't want to talk badly about prisma or the mongo integration. I think its a great achievement per se. 

But I strongly think something needs to be done immediatly. When fixing is not an option right now, beeing transparent about the caveats and limitations is the next best thing to do

**Reply by jbeckton** [reply]
Glad I came across this, I was just about to recommend Prisma for a MongoDB project at work. I figured it was ready for Production. I have not been using Mongo lately but rather Postgres and prisma has been great in that aspect.

**Reply by macrozone** [reply]
@jbeckton if your data set is small, it will be fine. Prisma is doing some wrong queries which do a COLSCAN on the database, which means each record needs to be loaded. On medium to large datasets, this can be very problematic.

For some there are workarounds (e..g you should never use findUnique or findUniqueOrThrow with mongodb as that has a severe bug)

On the other hand, mongodb is often not the perfect tool anyway for many applications. It has its uses, but if you need it for mostly relational data, use postgres

**Reply by mlmetzner** [reply]
yeah would generally avoid mongodb in the future, especially if you want to use prisma. one of my major pain points is that there is no way of doing a unique sparse index with prisma

**Comment by arvid220u** [comment]
We ran into similar problems with highly unexpected COLLSCANs caused by prisma-mongo in production.

Prisma is fantastic for Postgres, and I highly respect and admire the work that the Prisma team is doing. I think labeling prisma-mongo as GA runs the risk of tarnishing the reputation of the entire Prisma project, and would strongly recommend the team to move it back into "Early Access".

**Comment by macrozone** [comment]
seeing the recent activity in the repo for mongodb issues and even mrs gives me hope that prisma did not abandon mongodb!

Adding a test for colscans is an important step: https://github.com/prisma/prisma/pull/24071

**Comment by Jolg42** [comment, marked as answer]
Some good news 🙌🏼 

Our release from this week, 5.14.0, significantly improved the performance of certain queries when using MongoDB, specifically ones with `include` and `in` should now not cause a collection scan any more and use an index instead to find the correct documents. 

If you still encounter any performance problems with Prisma and MongoDB, please open a new GitHub bug issue. Thanks!

See release notes at https://github.com/prisma/prisma/releases/tag/5.14.0
